### PR TITLE
9FC02700.c: OK

### PR DESCRIPTION
**Pull request by pixel-stuck**

### CODE REVIEW

File: src/9FC02700.c [hunk 1]
```diff
@@ -1,27 +1,423 @@
...
+extern BbVirage01 D_9FC0F308;
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", verify_cert_chain);
+extern const char aRootCpca[];
+extern const char aRootXsca[];
+extern const char aRoot_0[];
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", get_expected_revocation_list_version);
+s32 cert_get_size(BbCertBase* certBase) {
+	switch(certBase->certType) {
```
Comment (Thar0) [inline src/9FC02700.c]:
```suggestion
    switch(certBase->certType) {
```

File: src/9FC02700.c [hunk 2]
```diff
@@ -1,27 +1,423 @@
...
+        return -1;
+    }
+
+    if (strcmp(toVerify->issuer, &aRoot) == 0) {
+        return rsa_verify_signature(&dataBlock, 1, pubkey, exponent, 1, signature);
+    } else if ((u32)toVerify->sigType < 2u) {
+        return rsa_verify_signature(&dataBlock, 1, toVeryifyAgainst->publicKey, toVeryifyAgainst->exponent, toVerify->sigType, signature);
+    } else {
+        return -1;
+    }
```
Comment (Thar0) [inline src/9FC02700.c]:
```suggestion
    }
    return -1;
```

File: src/9FC02700.c [hunk 3]
```diff
@@ -1,27 +1,423 @@
...
+}
+
+s32 verify_cert_chain(BbCertBase** certChain, s32 serverType) {
+    const char* serverName;
+    s32 i;
+
+    serverName = aXS;
+    if (serverType == 2) {
+        serverName = aCP;
+    }
```
Comment (Thar0) [inline src/9FC02700.c:95]:
```suggestion
    serverName = (serverType == 2) ? aCP : aXS;
```
Possibly?

File: src/9FC02700.c [hunk 4]
```diff
@@ -1,27 +1,423 @@
...
+            break;
+        case 1:
+            *out = &((BbRsaCert*)certBase)->signature;
+            break;
+        default:
+            return -1;
+    }
 
-INCLUDE_ASM("asm/non_matchings/9FC02700", verify_all_crlbundles);
+    return 0;
```
Comment (Thar0) [inline src/9FC02700.c:57]: Could this return be duplicated into case 0 and 1? Would make it a bit more like the others so maybe preferable idk

File: src/9FC02700.c [hunk 5]
```diff
@@ -1,27 +1,423 @@
...
+                return -1;
+        }
+
+        if (verify_cert_chain(bundle->certChain, t)) {
+            return -1;
+        }
+    }
+    
+    dataBlocks[0].data = &bundle->head->type;
+    dataBlocks[0].size = 0x58;
```
Comment (Thar0) [inline src/9FC02700.c]: Is this size expressible in terms of a `sizeof` of something?

File: src/9FC02700.c [hunk 6]
```diff
@@ -1,27 +1,423 @@
...
+    return 0;
+}
+
+s32 check_crlbundle_ranges(BbAppLaunchCrls* launchCrls) { // TODO: rename check_applaunch_crl_ranges
+    u32 i = 0;
+    BbCrlBundle* bundle;
+
+    for (i = 0, bundle = &launchCrls->tsrl; i < 3; i++, bundle++) {
+        if(bundle->head != NULL) {
+            if (check_untrusted_ptr_range(bundle->head, sizeof(BbCrlHead), 4) == 0) {
```
Comment (Thar0) [inline src/9FC02700.c]:
```suggestion
            if (!CHECK_UNTRUSTED(bundle->head)) {
```

File: src/9FC02700.c [hunk 7]
```diff
@@ -1,27 +1,423 @@
...
+                return 0;
+            }
+
+            if(check_untrusted_ptr_range(bundle->list, bundle->head->numberRevoked * sizeof(BbServerSuffix), 4) == 0) {
```
Comment (Thar0) [inline src/9FC02700.c:413]:
```suggestion
            if (!CHECK_UNTRUSTED_ARRAY(bundle->list, bundle->head->numberRevoked)) {
```
Not entirely sure about this one

File: src/9FC02700.c [hunk 8]
```diff
@@ -1,27 +1,423 @@
...
+                return 0;
+            }
+
+            if(bundle->head->type != 2) {
+                if (check_cert_ranges(bundle->certChain) == 0) {
+                    return 0;
+                }                
+            }
```
Comment (Thar0) [inline src/9FC02700.c]:
```suggestion
            if (bundle->head->type != 2 && check_cert_ranges(bundle->certChain) == 0) {
                return 0;
            }
```

File: src/9FC02700.c [hunk 9]
```diff
@@ -1,27 +1,423 @@
...
+        }
+    }
+
+    return 1;
+}
```
Comment (Thar0) [inline src/9FC02700.c]:
```suggestion
}

```

File: src/9FC02700.c [hunk 10]
```diff
@@ -1,27 +1,423 @@
...
+
+    return 0;
+}
+
+s32 check_certs_against_revocation_list(BbContentMetaDataHead* cmdHead, BbCertBase** chain, BbCrlBundle *arg2) {
+    s32 ret;
+    u32 i;
+
+    if (arg2[1].head != NULL) {
+        if (check_unknown_range(arg2[1].head, sizeof(BbCrlHead), 4) == 0) {
```
Comment (Thar0) [inline src/9FC02700.c]: These could do with a macro like `CHECK_UNTRUSTED`

File: src/9FC02700.c (hunk 5 above)
Comment (nim-ka) [reply src/9FC02700.c]: Yeah I have similar code in func_9FC00BAC

```c
    blocks[0].data = bundle->ticket;
    blocks[0].size = sizeof(BbTicket) - sizeof(BbRsaSig2048);
...
        blocks[0].data = &bundle->ticket->cmd.contentDesc;
        blocks[0].size = sizeof(BbContentDesc);
        blocks[1].data = &contentMetaDataHead;
        blocks[1].size = sizeof(BbContentMetaDataHead) - sizeof(BbRsaSig2048);
```

File: src/9FC02700.c (hunk 3 above)
Comment (pixel-stuck) [reply src/9FC02700.c:95]: unfortunately this doesn't match.

File: src/9FC02700.c [hunk 11]
```diff
@@ -1,27 +1,423 @@
 #include "include_asm.h"
```
Comment (nim-ka) [inline src/9FC02700.c]: can remove

File: src/9FC02700.c (hunk 4 above)
Comment (pixel-stuck) [reply src/9FC02700.c:57]: Unfortunately not. I also believe this is preferable, as most functions tend to have the success path at the bottom of the function like this, while the failure paths are early returns. Yeah, the above functions are different in that regard, though I think they're a bit different in terms of "success" vs "failure"

File: src/9FC02700.c (hunk 7 above)
Comment (pixel-stuck) [reply src/9FC02700.c:413]: yeah, unfortunately this one doesn't work. The size is variable, based on the number of revocations.

**Review by Thar0 — APPROVED** [review]

**Pull request merged**